node-odbc issues with Informix

srlopez opened the pull request:
I have an Ubutun 12.04 LTS 64 + unixODBC 2.3.1 + Informix clientsdk 4.10 + node.js v0.10.20
The Informix driver is correctly instaled because isql works ok.
But when i write the code in javascript and run width node, the trace seems to call **SQLDriverConnectW.c** and crash and the trace with isql call **SQLConnect.c** and works fine.

_partial code_

```
var Database = require("odbc").Database,
    db = new Database(),
    cn="DSN=NIFX"

db.open(cn, function (err) { //Crash
```

Here are the partial trace too
in NODE.JS

```
[ODBC][7171][1380810558.770556][__handles.c][460]
[ODBC][7171][1380810558.770647][SQLSetEnvAttr.c][189]
[ODBC][7171][1380810558.770677][SQLSetEnvAttr.c][363]
[ODBC][7171][1380810558.781819][SQLAllocHandle.c][375]
[ODBC][7171][1380810558.782068][SQLAllocHandle.c][493]
[ODBC][7171][1380810558.782802][SQLSetConnectAttrW.c][332]
[ODBC][7171][1380810558.782922][SQLSetConnectAttrW.c][616]
[ODBC][7171][1380810558.782968][SQLDriverConnectW.c][290]
        Entry:
            Connection = 0x7fc6000008c0
            Window Hdl = (nil)
            Str In = [DSN=NIFX][length = 9]
            Str Out = (nil)
            Str Out Max = 0
            Str Out Ptr = (nil)
            Completion = 0
        UNICODE Using encoding ASCII 'UTF8' and UNICODE 'UTF16LE'

[ODBC][7171][1380810558.787449][SQLDriverConnectW.c][699]
        Exit:[SQL_ERROR]
```

Running isql 

```
[ODBC][7151][1380810414.400667][__handles.c][460]
[ODBC][7151][1380810414.400743][SQLAllocHandle.c][375]
[ODBC][7151][1380810414.400773][SQLAllocHandle.c][493]
[ODBC][7151][1380810414.400803][SQLConnect.c][3700]
        Entry:
            Connection = 0x63d070
            Server Name = [NIFX][length = 4 (SQL_NTS)]
            User Name = [NULL]
            Authentication = [NULL]
        UNICODE Using encoding ASCII 'UTF8' and UNICODE 'UTF16LE'

[ODBC][7151][1380810414.803879][SQLConnect.c][4273]
        Exit:[SQL_SUCCESS]
```

Thanks in advance
_Santi_

platformamoja committed:
Added dynodbc module

wankdanker committed:
Add FETCH_ARRAY and FETCH_OBJECT constants

wankdanker committed:
ODBCConnection: accept Object as an argument to query functions and implment noResults flag to not create a result object

wankdanker committed:
ODBCStatement: initialize return values to supress compiler warnings

wankdanker committed:
ODBCResult: implement fetchMode option for fetch*() functions

wankdanker committed:
ODBCConnection: fix function arg index expectation

wankdanker committed:
optimize call to ODBCConnection::Query to make use of avoiding processing params

wankdanker committed:
implement ODBCConnection::CloseSync

wankdanker committed:
Update README.md to include new features

wankdanker committed:
Add openSync to js interace, Pool updates to avoid multiple ODBC objects

wankdanker committed:
Updated contributors

wankdanker committed:
Merge branch 'v0.5'

wankdanker committed:
Return more accurate column value types

This seems to be a pretty big speed-up; avoiding atoi() and atof().
Also, it is probably helpful to have floats as v8:Number and ints as
v8:Integer

wankdanker committed:
add benchmark for fetchAllSync

wankdanker committed:
increase query benchmark to 10000 iterations

wankdanker committed:
fix memory leak

wankdanker committed:
GetColumnValue optimization

wankdanker committed:
expose constants

wankdanker committed:
0.5.3

wankdanker committed:
Merge branch 'v0.5'

wankdanker committed:
whitespace changes to dynodbc code

wankdanker committed:
fixes for compiling dynodbc on linux

wankdanker committed:
expose dynodbc functionality to the public api

wankdanker committed:
include dynodbc in binding.gyp

wankdanker committed:
support setting library globally on the module

wankdanker committed:
remove duplicate pfnSQLBindParameter, dump dlopen() error messages

wankdanker committed:
only include sql header files if not using dynodbc

wankdanker committed:
add a couple library paths for testing

wankdanker committed:
0.5.4

wankdanker committed:
Merge branch 'v0.5'

wankdanker committed:
add @paulhendrix to contributors

wankdanker committed:
implement ODBCResult::FetchSync

wankdanker committed:
Avoid copying strings in synchronous methods

wankdanker committed:
Make sure result.closeSync returns true

wankdanker committed:
Implement ODBCStatement::ExecuteSync() and ExecuteDirectSync()

wankdanker committed:
Insure that ODBCStatement methods can be queued

wankdanker committed:
added test for statement executeSync

wankdanker committed:
Updated readme

wankdanker committed:
Fix internal handling of canFreeHandle on ODBCResult::New()

ODBCResult instances which are created from an ODBCStatement instance
should not be allowed to destroy themselves (or free the hSTMT to be
more specific) because the ODBCStatement can still be used to do stuff
with that hSTMT. Only ODBCResult instances that are created from an
ODBCConnection instance are allowed to actually destroy the hSTMT.

Previously, we were not properly derefencing the canFreeHandle flag
passed to ODBCResult::New() so it was evaluating to true in every case.
This was causing the ODBCResult::CloseSync() method to always destroy
the handle and thus cause the behaviour described in #32.

wankdanker committed:
Fix debug messages

wankdanker committed:
Test: drop tables before starting new test

wankdanker committed:
0.5.5

wankdanker committed:
Implement ODBCStatement::ExecuteNonQuery()

wankdanker committed:
Return number of rows affected from ODBCStatement::ExecuteNonQuery[Sync]

wankdanker committed:
Accept close options in ODBCStatement::CloseSync()

wankdanker committed:
test: log some information in test-describe-column

wankdanker committed:
ODBCResult: remove unuse variable

wankdanker committed:
0.5.6

wankdanker committed:
Allocate canFreeHandle on the heap before instantiating instances of ODBCResult

wankdanker committed:
Only free ODBCStatement parameters on Free(), Bind() and BindSync()

If parameters have been previously been bound, then the should be freed
before any new parameters are created. This would happen in Bind() or
BindSync(). They should also be freed in the dtor of the ODBCStatment
class.

This will allow us to bind once and repeatedly execte() and insert the
same values in case anyone needs to do this for some reason.

wankdanker committed:
Merge branch 'v0.5'

wankdanker committed:
Initialize timeInfo struct to avoid compiler warnings

wankdanker committed:
Properly return from CloseSync()

wankdanker committed:
Fix debug message

wankdanker committed:
Test: add test for rebinding values of parameter arrays

Not sure how or if this will work out

wankdanker committed:
Lock mutex when creating hENV

wankdanker committed:
Merge branch 'v0.5' of github.com:wankdanker/node-odbc into v0.5

wankdanker committed:
Lock mutex around SQLAllocHandle

wankdanker committed:
0.5.7

wankdanker committed:
Merge branch 'v0.5'

wankdanker committed:
More closely wrap handle allocs and frees with uv_mutex_lock

This is an ettempt to ensure that we are only blocking when we absolutely
have to. Previously we were blocking during the entire connect proccess.
This may speed some things up or it may break stuff. Revert this commit
if neccessary.

wankdanker committed:
Fix up remainging compiler warnings.

wankdanker committed:
0.5.8

wankdanker committed:
expose db.closeSync()

wankdanker committed:
Set fetchMode on result objects if fetchMode was passed to Database() constructor

wankdanker committed:
Add fetchMode property to ODBCResult and use as default

wankdanker committed:
Fix thread locking in ODBCConnection::Open[Sync]

wankdanker committed:
Updated tests and new tests

wankdanker committed:
Added benchmarks for comparing prepare vs not

wankdanker committed:
Updated benchmarks and start of separate bind queue

wankdanker committed:
Add next() function to SimpleQueue

wankdanker committed:
Only execute the next queued bind() call after an execute() call

wankdanker committed:
Add tests for bind/execute

wankdanker committed:
Accept params array to stmt.execute() and stmt.executeNonQuery

wankdanker committed:
Added new prepare benchmarks

wankdanker committed:
0.5.9

wankdanker committed:
After executeNonQuery[Sync] call SQLFreeStmt(..., SQL_CLOSE)

Hopefully this will avoid any invalid cursor state messages

wankdanker committed:
Fix executeNonQuery functino on constructor template

wankdanker committed:
test: executeSync() instead of async

wankdanker committed:
properly pass params to conn.query in queryResult

wankdanker committed:
fix SQLRowCount declaration in dynodbc

wankdanker committed:
0.5.10

wankdanker committed:
dynodbc: if dlsym fails, printf the function not loaded

wankdanker committed:
Update prefered way to execute a prepared statment

wankdanker committed:
Disable loading of SQLDrivers in dynodbc mode

wankdanker committed:
Attempt to determine if number is double or int in ODBC::GetParametersFromArray()

wankdanker committed:
test: update param test to test double vs int detection

wankdanker committed:
0.5.11

wankdanker committed:
Simplify number handling to just process as double

I am not entirely sure if there is any point in detecting if the
number is actually a double or an int. The overhead of detecting
which to use may outweigh that of using it. Will have to test
further, but if we want to go back to using the double/int
determination, then revert this commit

wankdanker committed:
Add ODBCResult::GetColumnNamesSync

This is so that we can retrieve the column names in case we are
fetching in a array mode.

wankdanker committed:
0.5.12

wankdanker committed:
Merge branch 'v0.5'

wankdanker committed:
Modify ODBC::GetSQLDiagRecError to take handleType and handle

wankdanker committed:
Replace SQLAlloc* with SQLAllocHandle

wankdanker committed:
Replace remaining SQLAllocStmt calls

wankdanker committed:
Proxy GetSQLError through GetDiagRec...

wankdanker committed:
Fix function signatures and add debug

wankdanker committed:
Properly loop through SQLGetDiagRec

wankdanker committed:
Check handle allocation succeeded

wankdanker committed:
Replace GetSQLDiagRecError with GetSQLError in all instances

What really happened is that I change the name of GetSQLDiagRecError
to GetSQLError and then modfied all calls to GetSQLDiagRecError and
GetSQLError so that they are calling properly. Also, GetSQLError is
now overloaded so that a custom error message can be added to the
error object, or not.
'

wankdanker committed:
0.5.13

wankdanker committed:
Merge branch 'v0.5'

wankdanker committed:
Merge branch 'master' of github.com:wankdanker/node-odbc

wankdanker committed:
Modify dynodbc to not load functions we don't use

wankdanker committed:
Replace SQLSetConnectOption with SQLSetConnectAttr; closes #38

wankdanker committed:
Add libsqlncli library to test with

wankdanker committed:
0.5.14

wankdanker committed:
Merge branch 'v0.5'

wankdanker committed:
Fix windows time struct declaration

wankdanker committed:
0.5.15

wankdanker committed:
Merge branch 'v0.5'

wankdanker committed:
Try to fix windows unicode support by using SQL_C_WCHAR

Also have to create the return string in a special way with some type casting

wankdanker committed:
0.5.16

wankdanker committed:
Merge branch 'v0.5'

wankdanker committed:
Unicode progress; all tests work now except unicode. DOH.

wankdanker committed:
More Unicode updates

wankdanker committed:
Param set type to SQL_WVARHCAR, debug info

wankdanker committed:
Use Unicode column retrieval for linux also.

wankdanker committed:
Cast to SQLWCHAR

wankdanker committed:
Use Local<String> and Write instead of memcopy

wankdanker committed:
New and updated unicode tests

wankdanker committed:
Modify GetParametersFromArray to use string->Write instead of memcpy

wankdanker committed:
Use sql as void* and cast where necessary

wankdanker committed:
Update characters included in the unicode test

wankdanker committed:
0.5.17

wankdanker committed:
Merge branch 'v0.5'

wankdanker committed:
Add more REQ_STR for non-Utf8Value

wankdanker committed:
Enable Unicode connection strings

wankdanker committed:
unicode: start properly handling Unicode based on defined UNICODE flag

wankdanker committed:
unicode: finish up proper handling of Unicode in all classes

wankdanker committed:
fix: use SQL_DESC_TYPE to get the column type

wankdanker committed:
fix: re-introduce connectionString objects

wankdanker committed:
test: update test runner to test multiple drivers

wankdanker committed:
fix: specify SQL_WVARCHAR or SQL_VARCHAR for params depending on UNICODE

wankdanker committed:
test: updated tests and new default for MSSQL Native Client connectionString

wankdanker committed:
bench: move insert benchmark to disabled

wankdanker committed:
bench: update benchmarks to run with run-bench

Eventually we will have a table which displays results for each driver

wankdanker committed:
Merge branch 'unicode-madness' into v0.5

wankdanker committed:
0.5.18

wankdanker committed:
Merge branch 'v0.5'

wankdanker committed:
fix: malloc error

wankdanker committed:
debug: update debug messages

wankdanker committed:
Specify DecimalDigits when binding param"

wankdanker committed:
fix-attempt: in the case of type double binding params, set decimals to 7.

This is probably a terrible idea, but so far MSSQL drivers don't work
any other way. While MySQL and Sqlite seem to ignore this value because
it does not effect the precision of the input.

wankdanker committed:
0.5.19

wankdanker committed:
Merge branch 'v0.5'

wankdanker committed:
Add STRICT_COLUMN_NAMES compile option and documentation.

wankdanker committed:
0.5.20

wankdanker committed:
Disable OutConnectionString in SQLDriverConnect

We weren't using it anyway and it was an extra malloc

wankdanker committed:
Use Local<Object> instead of Persisten<Object> for ODBC* instances

wankdanker committed:
Use proper HDBC type when getting js_hdbc->Value()

wankdanker committed:
Set connected status after closeSync()

wankdanker committed:
statement: free SQL_C_WCHAR param types

wankdanker committed:
statement: throw errors on prepareSync

wankdanker committed:
test: add test to try to invoke errors during prepare/bind/execute

wankdanker committed:
Set the prototype of our error objects to Exception::Error

This allows node's default error handling to display stack
traces and such. So much nicer than just [Object object]

wankdanker committed:
Don't set message manually on error object, use Exception::Error constructor for that

wankdanker committed:
test: update test to assert throw/doesNotThrow

wankdanker committed:
0.5.21

wankdanker committed:
Use SQL_LONGVARCHAR for string parameters

wankdanker committed:
Add test for binding a long string

wankdanker committed:
0.5.22

wankdanker committed:
GetColumnValue: check if len == SQL_NULL_DATA instead of ret

wankdanker committed:
GetColumnValue: add support for returning really long string column values

wankdanker committed:
GetSQLError: re-introduce the explicit message property

wankdanker committed:
test: udpate some tests

wankdanker committed:
test: reduce number of iterations; causing false failure due to timeout

wankdanker committed:
GetColumnValue: remove comparison of len < 0

That was just an alternate way of checking len == SQL_NULL_DATA

wankdanker committed:
test: add note to test

wankdanker committed:
GetColumnValue: use handle scope

wankdanker committed:
0.5.23

wankdanker committed:
test: add test for domains

wankdanker committed:
test: enclose column names in double quotes so they are case sensitive in postgres

wankdanker committed:
Merge branch 'v0.5' of github.com:wankdanker/node-odbc into v0.5

wankdanker committed:
Add support for setting connectTimeout. Closes #50.

Set the connectTimeout at runtime by passing and options object to the Database
constructor function:

var db = require('odbc')({ connectTimeout : 60 });

or

var odbc = require('odbc');
var db = new odbc.Database({ connectTimeout : 42 });

wankdanker committed:
0.5.24

wankdanker committed:
Allow connectTimeout of zero

wankdanker committed:
Conditionally use SQL_[W]LONGVARCHAR or SQL_[W]VARCHAR based on length of parameter

wankdanker committed:
Make sure async callbacks are wrapped in TryCatch

wankdanker committed:
test: replace open() and close() calls with openSync() and closeSync()

Only made the replacement in places where we are not explicitly
testing the open() or close() functionality. The reason behind
this is that MySQL has a bug which causes most of the tests to
time out because a thread is not released if a connection is
opened and closed from two different threads. This tends to occur
when we are using the async calls because the ODBC open/close call
occurs on the thread pool. And we can't guarantee that it will
be on the same thread. The only way to guarantee ODBC open/close
is called from the same thread is if we call it from the main
thread. In order to do that, we must use the Sync versions of
open() or close()

wankdanker committed:
0.5.25

wankdanker committed:
Check to see if 'cb' exists in 'close()' before calling it

theduderog committed:
Attempt to handle INVALID_HANDLE errors

theduderog committed:
Added some comments

wankdanker committed:
Merge branch 'infinite_loop' of https://github.com/theduderog/node-odbc into test-merge

wankdanker committed:
Add test for pull request issue

wankdanker committed:
Close result objects in columns and tables proxy functions

wankdanker committed:
Add a debug message

wankdanker committed:
Add an assertion that return value from SQLGetData must not be -2

wankdanker committed:
Updated most recent issue test

wankdanker committed:
Merge branch 'v0.5' into test-merge

wankdanker committed:
Move assertion to inside error handling and attempt to throw an error object

This could be bad because are in some cases expecting to pass
the error object back as an argument to the callback function.
Other times it might be ok to actually throw. At any case, this
should be a super rare error to hit.

wankdanker committed:
0.5.26

wankdanker committed:
Add SQLGetDiagField  to dynodbc - untested

wankdanker committed:
0.5.27

wankdanker committed:
Make dynodbc.h compile in dynodbc mode with SQLTCHARS and other such modifications'

wankdanker committed:
0.5.28

wankdanker committed:
Add new test for #53

wankdanker committed:
Add an open method which returns a new Database instance and an open connection

wankdanker committed:
0.5.29

wankdanker committed:
Add local path for macs in binding.gyp. Fixes #52

wankdanker committed:
0.5.30

lee-houghton committed:
Handle errors returned by SQLMoreResults

lee-houghton committed:
Handle error returned by moreResultsSync()

Handle error returned by moreResultsSync() in Database.query. Return the
error as a separate result set.

wankdanker committed:
Fixup pull request for issue #54

Added test for the issue which is lacking assertions
ThrowException instead of returning the error object
Keep looping and emitting errors if encountered when calling moreResultsSync()

wankdanker committed:
0.5.31

wankdanker committed:
Update package.json to point to wankdanker's repo

Sorry Lee. Hope you don't mind. :)

lee-houghton committed:
Fix blank result in ODBC::GetSQLError

It seems with MSSQL at least, SQLGetDiagField needs to be called with
record number 0 when retrieving header fields such as the error count
(even though the documentation states that the record number is
ignored). This stops SQLGetDiagField returning SQL _ERROR and thus
numfields being uninitialized.

lee-houghton committed:
More fixes for errors and multiple result sets

1. If the first result set is an error, still construct a result object
as there may be further result sets, e.g.
```SQL
select 1/0 as [infinity];
select 'This message should be returned' as [message];
```

2. Secondly, fix a case where a spurious result set is produced
following multiple errors:
```SQL
select 1 as i;
raiserror('Error #1', 16, 1);
raiserror('Error #2', 16, 1);
select 'You should not see a blank result set before this message' as
[message];
```

wankdanker commented:
Not sure if you meant this to be a pull request or just an issue....

Have you tried compiling without UNICODE enabled? https://github.com/wankdanker/node-odbc#unicode

I wonder if that would change anything.

Dan

lee-houghton committed:
Fix undefined behaviour in UV_AfterExecute

args is declared as "Local<Value> args[3];" but then args[3] is
assigned...

lee-houghton committed:
More arrays declared wrong

wankdanker committed:
Cast error string to avoid compiler warning

wankdanker committed:
0.5.32

wankdanker committed:
Finish implementing transactions and update docs.

wankdanker committed:
0.6.0

wankdanker committed:
Document additional way to create an instance of Database

wankdanker committed:
Update issues url

wankdanker committed:
0.6.1

wankdanker committed:
docs: remove callback param from Sync functions; silly copy and paste.

wankdanker committed:
Change unix time conversion to use timelocal instead of mktime

We noticed that in the normal usage of the module, our times were reported
incorrectly. It traced back to using timelocal or timegm. timelocal works
as we expect in more cases. There is now a define available if you need
it to use timegm (see README.md)

wankdanker committed:
0.6.2

Commit:
Update README.md

SERVERNAME instead of SERVER

wankdanker committed:
Fix error message buffer size bug

We were passing the size of the message buffer in bytes
instead of the number of characters. This caused issues
when UNICODE is being used because there are at least 2
bytes per character. This was causing buffer over-runs.

Instead of setting the buffer size with a literl number
I have used the constants ERROR_MESSAGE_BUFFER_BYTES and
ERROR_MESSAGE_BUFFER_CHARS

Closes #57

wankdanker committed:
0.6.3

rossipedia committed:
Fix for segfault 11 on OSX

Probably not a good idea to access memory after it's been freed

wankdanker committed:
0.6.4

rossipedia committed:
Use C99 designated struct initializer (removes warnings on OSX)

wankdanker committed:
Fix connectTimeout and add loginTimeout

This should fix #50. Previously we were using SQL_ATTR_LOGIN_TIMEOUT as the connect
attribute for connectionTimeout. I think this is wrong. I have replaced the attribute
with SQL_ATTR_CONNECTION_TIMEOUT and introduce another attribute called loginTimeout
to set SQL_ATTR_LOGIN_TIMEOUT.

wankdanker committed:
Properly accept 0 as a value for connectTimeout and loginTimeout

wankdanker committed:
Set default connectTimeout to 0

This in effect preserves the previous behavior where we would
set SQL_ATTR_LOGIN_TIMEOUT to 5. It seems that setting a timeout
value for SQL_ATTR_CONNECTION_TIMEOUT causes an assertion error
with the FreeTDS driver

wankdanker committed:
Use SQLUINTEGER for connectTimeout and loginTimeout

Before we were using plain int. This is just an attempt to be accurate
in hopes of fixing a FreeTDS assertion when connecting with a value for
connectTimeout

wankdanker committed:
Update test-binding-connection-timeOut to reflect the new default value of connectTimeout

wankdanker committed:
Use member value of connectTimeout or loginTimeout directly

Instead of making a local copy of connectTimeout and loginTimeout
use the value directly on the object like: &self->connectTimeout

I am weary of doing this because I've had issues in the past with
creating a pointer from a member variable and it is just my own
lack of knowledge of C/C++ that makes me feel this way. However,
in use this way it does not cause an error assertion in FreeTDS

wankdanker committed:
Readability for getting the address of members

wankdanker committed:
Merge branch 'v0.6' into fix-connectTimeout

wankdanker committed:
Call SQLSetConnectAttr properly

Evidently we must pass the value of the connectTimeout or loginTimeout to
SQLSetConnectAttr, not a pointer to the value. This is super counter-intuitive
because we have to cast the value to a SQLPOINTER in order to get it to even
compile. In order to avoid compiler warnings we must cast to size_t before
casting to SQLPOINTER.

This will properly timeout connections when attempting to connect to a
non-running server; at least with the FreeTDS driver.

wankdanker committed:
Prevent infinite loop when processing string values

A corner case occurred which was probably caused by a ODBC driver behaving
incorrectly. The first call to SQLGetData would return len == 0 which
indicates an empty string. The next time through the loop we would expect
SQLGetData to return ret == SQL_NO_DATA. This is not happening in some
cases, so an infinite loop occurs.

To fix this we check if len == 0 before looping again, if so we break out
of the loop. This is a slight optimization because if we know len == 0 then
there is no point in calling SQLGetData again to see if there are more data.

wankdanker committed:
Merge branch 'fix-connectTimeout' into v0.6

wankdanker committed:
0.6.5

wankdanker committed:
Revert "Use C99 designated struct initializer (removes warnings on OSX)"

This reverts commit 3cd950154cc74ac5d8117fbe7ddf9e9fb0212bfd.

wankdanker committed:
0.6.6

wankdanker committed:
Enable setting options in the pool

wankdanker committed:
Double check that `str` has some data to avoid returning an empty handle

wankdanker committed:
Add debug message to ODBCResult destructor

wankdanker committed:
Queue the ODBCResult.fetch() method

wankdanker committed:
the white spaces

wankdanker committed:
Create a default error message when no diagnostic information available

wankdanker committed:
Attempt to use fwprintf when using DEBUG_PRINTF with UNICODE enabled

wankdanker committed:
Don't return Null() when len == SQL_NULL_DATA if we have str data

Fixes #60

wankdanker committed:
0.6.7

wankdanker commented:
@srlopez Please open a new issue (non-pull request) if this is still an issue for you.

Thanks.
